tests/chainer_tests/training_tests/updaters_tests/test_standard_updater.py

```diff
@@ -115,6 +118,257 @@ def test_serialize(self):
+        model.to_device(model_initial_device)
+        optimizer = DummyOptimizer()
+        optimizer.setup(model)
+        iterator = DummyIterator([numpy.array(1), numpy.array(2)])
+
+        updater = training.updaters.StandardUpdater(
+            iterator,
+            optimizer,
+            device=device,
+            input_device=input_device)
+
+        assert updater.device is device
+        assert updater.input_device is input_device
+
+        # Check the model device.
+        assert model.device == device
+
+        updater.update_core()
+
+        assert optimizer.update.call_count == 1
+        args, kwargs = optimizer.update.call_args
+        assert len(args) == 2
+        assert len(kwargs) == 0
+        loss, v1 = args
+
+        # Check the input device.
+        assert backend.get_device_from_array(v1) == input_device
+
+
+class DummyDevice(backend.Device):
+
+    xp = numpy
+    supported_array_types = (numpy.ndarray,)
+
+    def __init__(self, index):
+        self.index = index
+
+    def __eq__(self, other):
+        return isinstance(other, DummyDevice) and other.index == self.index
+
+    # TODO(niboshi): Define name property instead (#7149).
+    def __str__(self):
+        return '@dummy:{}'.format(self.index)
+
+    def send_array(self, array):
+        return array.copy()
+
+
+@testing.parameterize(*testing.product({
+    'omit_device': [True, False],
+    'omit_input_device': [True, False],
+}))
+class TestStandardUpdaterDeviceArgumentFallback(unittest.TestCase):
+    """Tests the fallback behavior regarding device and input_device
+    arguments."""
+
+    def test_device_argument_fallback(self):
+        self.check_device_argument_fallback(
+            initial_model_device=DummyDevice(0),
+            initial_input_device=DummyDevice(1),
+            device_arg=DummyDevice(3),
+            input_device_arg=DummyDevice(4))
+
+    @attr.multi_gpu(2)
+    def test_gpu_to_gpu_transfer(self):
+        # GpuDevice is given as device arguments:
+        # - model GPU-to-GPU transfer should be skipped.
+        # - input GPU-to-GPU transfer should NOT be skipped.
+        self.check_device_argument_fallback(
+            initial_model_device=backend.GpuDevice.from_device_id(0),
+            initial_input_device=backend.GpuDevice.from_device_id(0),
+            device_arg=backend.GpuDevice.from_device_id(1),
+            input_device_arg=backend.GpuDevice.from_device_id(1))
+
+    def _get_expected_devices(
+            self,
+            initial_model_device,
+            initial_input_device,
+            device_arg,
+            input_device_arg):
+        # Determines the expected devices.
+        # Returns: (
+        #    expected_device_attr:   Expected StandardUpdater.device
+        #    expected_model_device:  Expected model device
+        #    expected_input_device:  Expected device given to converters
+        # )
+        # or None if an error is expected.
+
+        if device_arg is not None:
```

Comment:
Do you need this condition (you can reduce one indentation level and also think it's easier to read)?

Reply:
Fixed.